core/src/main/java/org/apache/druid/guice/GuiceAnnotationIntrospector.java

```diff
@@ -58,9 +58,9 @@ public Object findInjectableValueId(AnnotatedMember m)
       if (m instanceof AnnotatedMethod) {
         throw new IAE("Annotated methods don't work very well yet...");
       }
-      return Key.get(m.getGenericType());
+      return Key.get(m.getType());
     }
-    return Key.get(m.getGenericType(), guiceAnnotation);
+    return Key.get(m.getType(), guiceAnnotation);
   }
 
```

Comment:
```suggestion
    return Key.get(getParamType(m), guiceAnnotation);
  }

  private Type getParamType(AnnotatedMember m) {
    if(m.getType().isContainerType()){
      return Types.newParameterizedType(m.getType().getRawClass(), m.getType().getContentType().getRawClass());
    }
    return  m.getRawType();
  }
```

This seems to work for me (and of course changing the Key.get(m.getType()); to Key.get(getParamType(m)); on line 61.

But this was on version 0.22.1 and I'm not sure if this is as robust handling that could be needed.

Reply:
Thanks for the suggestion, @CookieAroundTheBend ! I will try to spend some time on this and get it resolved soon.